pkg/standalone-utils/contracts/PermitRelayer.sol

```diff
@@ -0,0 +1,149 @@
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20PermitDAI.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+import "./base/MultiDelegatecall.sol";
+
+/**
+ * @title PermitRelayer
+ * @notice Allows users to atomically perform multiple Balancer Vault actions in sequence
+ * including token approvals using permit (where supported)
+ */
+contract PermitRelayer is MultiDelegatecall {
+    IVault public immutable vault;
+
+    constructor(IVault _vault) {
+        vault = _vault;
+    }
+
+    /**
+     * @notice Allows calling an arbitrary function on the Vault
+     */
+    function _vaultAction(uint256 value, bytes memory data) private returns (bytes memory) {
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory result) = address(vault).call{ value: value }(data);
+
+        // Pass up revert if the call failed
+        if (!success) {
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
+            }
+        }
+
+        return result;
+    }
+
+    function setRelayerApproval(
+        address relayer,
+        bool approved,
+        bytes calldata authorisation
+    ) external payable {
+        bytes memory data = abi.encodePacked(
+            abi.encodeWithSelector(vault.setRelayerApproval.selector, msg.sender, relayer, approved),
+            authorisation
+        );
+        _vaultAction(0, data);
+    }
+
+    function swap(
+        IVault.SingleSwap calldata singleSwap,
+        IVault.FundManagement calldata funds,
+        uint256 limit,
+        uint256 deadline,
+        uint256 value
+    ) external payable returns (uint256) {
+        require(funds.sender == msg.sender, "Incorrect sender");
+        return vault.swap{ value: value }(singleSwap, funds, limit, deadline);
+    }
+
+    function batchSwap(
+        IVault.SwapKind kind,
+        IVault.BatchSwapStep[] calldata swaps,
+        IAsset[] calldata assets,
+        IVault.FundManagement calldata funds,
+        int256[] calldata limits,
+        uint256 deadline,
+        uint256 value
+    ) external payable returns (int256[] memory) {
+        require(funds.sender == msg.sender, "Incorrect sender");
+        return vault.batchSwap{ value: value }(kind, swaps, assets, funds, limits, deadline);
+    }
+
+    function manageUserBalance(IVault.UserBalanceOp[] calldata ops, uint256 value) external payable {
+        for (uint256 i = 0; i < ops.length; i++) {
+            require(ops[i].sender == msg.sender, "Incorrect sender");
+        }
+        vault.manageUserBalance{ value: value }(ops);
+    }
+
+    function joinPool(
+        bytes32 poolId,
+        address recipient,
+        IVault.JoinPoolRequest calldata request,
+        uint256 value
+    ) external payable {
+        vault.joinPool{ value: value }(poolId, msg.sender, recipient, request);
```

Comment:
I'd receive `sender` and check that it matches `msg.sender`, to be consistent with the other functions.